Contribution documentation + PR template. (#19)
fewlinesco/connect-account: README.adoc

= Connect Account

> Connect Account is an open source app, which aims at providing all the business logic to handle connect identities and profile data, and a default frontend for user to use it.

Reviewer: ```suggestion
Connect Account is an open source app, which aims at providing all the business logic to handle Connect identities and profile data, and a default frontend for Users to use it.
```

You will most likely need to fork the repository to add your custom frontend component, to change the layout, or add your branding and design-system.

‼️You'll need to manually update your **origin** master branch each time before making a new branch from your fork. To do so, run the following, from your **local** master:

[source, shell]
----
git fetch --upstream
git rebase --upstream
git push origin master
git checkout -b name-of-your-branch
----

== Forking

> 
> 

Reviewer: ```suggestion

```

= Connect Account
:toc: preamble

Connect Account is an open source app, which aims at providing all the business logic to handle Connect identities and profile data, and a default frontend for User to use it.

== Flows

=== Connect authentication flow

image::./doc/oauth2-flow-diagram.svg[OAuth2 flow diagram]

> For more information, check the Authorization code flow https://developers.fewlines.co/guides/authorization_code_flow/overview/[documention].

Reviewer 2: ```suggestion
For more information, check the Authorization code flow https://developers.fewlines.co/guides/authorization_code_flow/overview/[documentation].
```

Reviewer 2: Please, don't tell me that this typo is the real URL. 😅

Author: Haha no, it's just the way `asciidoc` handle hyperlinks.

Author: a62d1f2 Fixed. If you could re-approved, thanks :).

Get required environment variables by copying `.env.sample` and if necessary tuning it to values relevant to your dev environment:

[source, shell]
----
cp .env_vars.sample .env_vars
----

Source the file to get variables defined your environnement:

[source, shell]
----
source .env_vars
----

== Development Process

[source, shell]
----

> NODE_TLS_REJECT_UNAUTHORIZED=0 yarn dev

Reviewer: ```suggestion
yarn dev

# or, if you're working with a local version of Connect
NODE_TLS_REJECT_UNAUTHORIZED=0 yarn dev
```

| URL used to fetch identities from the management GraphQL endpoint.

| `CONNECT_MANAGEMENT_API_KEY`
| API key used to access the management GraphQL endpoint.

| `CONNECT_PROVIDER_URL`
| URL used to start the connect oauth flow.

| `CONNECT_APPLICATION_CLIENT_ID`
| Client ID of the online service (e.g. internet website, application) that uses the Provider Authentication and Authorization service for its User.

| `CONNECT_APPLICATION_CLIENT_SECRET`
| Paired with the client ID, used to authenticate the Application from which the User intent to sign in.

| `CONNECT_APPLICATION_SCOPES`
| Represents the kind of user authorized information and actions that an Application is able to access on another Application.

| `NEXT_PUBLIC_SENTRY_DSN`
| Data Source Name representing the configuration required by the Sentry SDKs.
|===

== Connect login implementation

To understand the flow of `connect-account`, you should read the `connect` https://developers.fewlines.co/guides/authorization_code_flow/overview/[documentation].

[source, jsx]
----
<BusinessComponent>
  {({ businessLogic }) => (
    <VisualizationComponent businessLogic={businessLogic} />
  )}
</BusinessComponent>

> ----

Reviewer: Shouldn't this be in the `=== Splitting logic from visualization components` section?

Author: Nice catch, dunno why it went there in the end ahah.

== Source Code Organization

We are using the NextJS folder architecture (i.e. `/pages`) to utilize its router, out of the box. For more information, please refer to the https://nextjs.org/docs/basic-features/pages[documentation].

We are also using the `Command Query Responsibility Segregation`(CQRS) pattern to separate queries from mutations. They are located in the `queries/` and `command/` folder.

=== /.github

- */workflows*: GitHub Actions used to run tests during CI/CD process flow.
- *PULL_REQUEST_TEMPLATE*: Template used when opening a pull request on GitHub.

=== /src

- *@types/*: Type declaration used in multiple places.
- *assets/*: Folder containing assets.
- *commands/*: Write (e.g. `POST`) database actions.
- *components/*:

>   - *business/*: React components that provide business logic through render props. You shouldn't touch them.

Reviewer: ```suggestion
  - *business/*: React components that provide business logic through render props. You shouldn't have to touch them.
```

>   - *visualization/*: React component that  

Reviewer: that?

Author: > React component that only — for the most part — handle the visualization part. 

Not a huge fan but I can't really find better wording.

Reviewer: 👍

    - *fewlines/*: Our default visualization components. Used as a template.
- *design-system/*: Style or atomic components (e.g. icons) related code.
- *hooks/*: Reusable actions flows.
- *middleware/*: Reusable wrappers to add various features to server side actions.
- *pages/*: NextJS router.
- *queries/*: Read (i.e. `GET`) database actions.
- *utils/*: Small snippets/functions used multiple times throughout the application
- *config.ts*: Entry point used to verify env vars sourcing, and prevent the app to run if forgotten.

=== /tests
- *__mocks__*: Centralized mocked data or function.
- **.test.(ts|tsx)*: Unit test files.

=== config files
- *package.json*: We use this file, as much as possible as a centralized config file for various packages, like `ESLint`, `Jest` or `Husky`.
- *tsconfig.json*: TypeScript compiler options.
- *next.config.js*: Extended webpack compiler config used by NextJS.

> === Splitting logic from visualization components

Reviewer: ```suggestion
=== Spliting logic from visualization components
```

Author: [:D](https://www.linguee.fr/anglais-francais/traduction/splitting.html)

> To help minimized conflict from upstream in your fork, we chose to split the business logic — which is handle by us — from visualization components. To do so, we are using https://reactjs.org/docs/render-props.html#gatsby-focus-wrapper[Render Props]:

Reviewer: ```suggestion
To help minimized conflict from upstream in your fork, we chose to split the business logic — which is handled by us — from visualization components. To do so, we are using https://reactjs.org/docs/render-props.html#gatsby-focus-wrapper[Render Props]:
```

Reviewer: ```suggestion
To help minimized conflict from upstream in your fork, we chose to split the business logic — which is handle by us — from visualization components. To do so, we are using https://reactjs.org/docs/render-props.html[Render Props]:
```

or am I missing something?

Author: Woooops

The `globalStyle` object, found in `/src/design-system/globals/globalStyle.tsx`, is used to remove undesired style and behavior found in HTML. 

> Note that we chose to set the global font size to *62.5%*. This font size means that '1rem' is exactly equal to '10px', which makes setting REM values very simple.

The `globalStyle` is scoped in the `_app` page component.

=== Theme

You should prioritize theme values over arbitrary CSS values.

The theme structure is defined inside the `src/@types/styled-component.d.ts` definition type file.

The theme values are set inside `src/design-system/theme/lightTheme.ts` file. Each field are set individually, then assigned to the `lightTheme` variable.

The theme access is scoped in the `_app` page component.

You can access the theme properties and values inside a `styled` component by calling the `theme` props:

[source, typescript]
----
import styled from "styled-components";

// Note that styled component are PascalCase.
const StyledComponent = styled.div`
  css-property: ${({ theme }) => theme.themeKey};
`
----

You can access the theme properties and values from everywhere else with the custom `useTheme` hook:

[source, typescript]
----

> import { useTheme } from "relative/path/to/the/hook";

Reviewer: ```suggestion
import { useTheme } from "relative/path/to/the/hooks/useTheme";
```

where is it by the way?

Author: `src/design-system/theme/useTheme.ts`

Author: I'm still wondering if it should go into `/hooks` though.

Reviewer: ```suggestion
import { useTheme } from "relative/path/to/the/design-system/theme/useTheme";
```
then 😛
